src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp

```diff
@@ -218,34 +218,29 @@ void ShenandoahGenerationalHeap::compute_old_generation_balance(size_t old_xfer_
   assert(old_reserve <= max_old_reserve, "cannot reserve more than max for old evacuations");
 
   // We now check if the old generation is running a surplus or a deficit.
-  size_t old_region_deficit = 0;
-  size_t old_region_surplus = 0;
-
   const size_t max_old_available = old_generation()->available() + old_cset_regions * region_size_bytes;
   if (max_old_available >= old_reserve) {
     // We are running a surplus, so the old region surplus can go to young
     const size_t old_surplus = max_old_available - old_reserve;
-    old_region_surplus = old_surplus / region_size_bytes;
     const size_t unaffiliated_old_regions = old_generation()->free_unaffiliated_regions() + old_cset_regions;
-    old_region_surplus = MIN2(old_region_surplus, unaffiliated_old_regions);
+    const size_t old_region_surplus = MIN2(old_surplus / region_size_bytes, unaffiliated_old_regions);
+    old_generation()->set_region_balance(checked_cast<ssize_t>(old_region_surplus));
   } else {
     // We are running a deficit which we'd like to fill from young.
     // Ignore that this will directly impact young_generation()->max_capacity(),
     // indirectly impacting young_reserve and old_reserve.  These computations are conservative.
     const size_t old_need = old_reserve - max_old_available;
+    const size_t max_old_region_xfer = old_xfer_limit / region_size_bytes;
+
     // The old region deficit (rounded up) will come from young
-    old_region_deficit = (old_need + region_size_bytes - 1) / region_size_bytes;
+    size_t old_region_deficit = (old_need + region_size_bytes - 1) / region_size_bytes;
 
     // Round down the regions we can transfer from young to old. If we're running short
     // on young-gen memory, we restrict the xfer. Old-gen collection activities will be
     // curtailed if the budget is restricted.
-    const size_t max_old_region_xfer = old_xfer_limit / region_size_bytes;
     old_region_deficit = MIN2(old_region_deficit, max_old_region_xfer);
+    old_generation()->set_region_balance(0 - checked_cast<ssize_t>(old_region_deficit));
```

Comment:
The code in the if and the else branches is highly isomorphic. I have the feeling you may be able to fold them together, but I am not sure how clean it will be. Also if you start with `ssize_t` for some of the operands that feed into the surplus calculation (or balance if you will, but we can call it surplus wit the understanding that a negative surplus is a deficit), then you may be able to avoid some of the later casts and the signing of unsigned quantities, including the `(cast)x`, `-(cast)x` or `0-(cast)x`constructs above.

Reply:
I agree the code in the branches is similar, but the APIs we have for managing capacity for a generation will ultimately make us chose old or young for the source or destination of the transfer based on whether `old_region_balance` is positive or negative.

Reply:
Yes, I was thinking that the generation sizer could do that based on the sign of the balance.
But I realize that might not necessarily simplify the code, as you seem to have discovered. OK, thanks for checking.